cdap-app-fabric/src/main/java/io/cdap/cdap/internal/sysapp/SystemAppManagementService.java

```diff
@@ -0,0 +1,135 @@
+import com.google.gson.JsonParseException;
+import com.google.inject.Inject;
+import io.cdap.cdap.common.conf.CConfiguration;
+import io.cdap.cdap.common.conf.Constants;
+import io.cdap.cdap.internal.app.services.ApplicationLifecycleService;
+import io.cdap.cdap.internal.app.services.ProgramLifecycleService;
+import org.apache.twill.common.Threads;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.IOException;
+import java.io.Reader;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+
+/**
+ * Enables system apps using system app config files located in SYSTEM_APP_CONFIG_DIR. Each config
+ * describes steps to perform to start one (or more) system app. Currently
+ * SystemAppManagementService runs during bootstrap phase.
+ */
+public class SystemAppManagementService extends AbstractIdleService {
+
+  private static final Logger LOG = LoggerFactory.getLogger(SystemAppManagementService.class);
+  private static final Gson GSON = new Gson();
+
+  private final File systemAppConfigDirPath;
+  private final SystemAppEnableExecutor systemAppEnableExecutor;
+  private ExecutorService executorService;
+
+  @Inject
+  SystemAppManagementService(CConfiguration cConf, ApplicationLifecycleService appLifecycleService,
+                             ProgramLifecycleService programLifecycleService) {
+    this.systemAppConfigDirPath = new File(cConf.get(Constants.SYSTEM_APP_CONFIG_DIR));
+    this.systemAppEnableExecutor = new SystemAppEnableExecutor(appLifecycleService,
+        programLifecycleService);
+  }
+
+  @Override
+  protected void startUp() {
+    LOG.debug("Starting {}", getClass().getSimpleName());
+    executorService = Executors
+        .newSingleThreadExecutor(Threads.createDaemonThreadFactory("sys-app-management-service"));
+    executorService.execute(() -> {
+      try {
+        // Execute all steps for each config file in system config directory.
+        bootStrapSystemAppConfigDir();
+      } catch (Exception ex) {
+        LOG.error("Got exception in watch service for system app config dir", ex);
+      }
+    });
+  }
+
+  private void bootStrapSystemAppConfigDir() throws Exception {
+    LOG.debug("Number of config files {} in system app config directory.",
+        systemAppConfigDirPath.listFiles().length);
+
+    for (File sysAppConfigFile : systemAppConfigDirPath.listFiles()) {
+      LOG.debug("Running steps in config file {}", sysAppConfigFile.getAbsoluteFile());
+      executeSysConfig(sysAppConfigFile.getAbsoluteFile());
+    }
+  }
+
+  private void executeSysConfig(File fileName) throws Exception {
+    SystemAppConfig config = parseConfig(fileName);
+    for (SystemAppStep step : config.getSteps()) {
+      try {
+        step.validate();
+      } catch (IllegalArgumentException e) {
+        LOG.warn("Config step {} failed because it is malformed: {}", step.getLabel(),
```

Comment:
Log with the exception, not just the `e.getMessage`. Otherwise the stacktrace will be lost and hard to find the root cause.